alacritty_terminal/src/term/cell.rs

```diff
@@ -113,8 +126,33 @@ impl Cell {
         self.drop_extra();
         self.c = ' ';
     }
+
+    /// Set underline color on the cell.
+    pub(crate) fn set_underline_color(&mut self, color: Option<Color>) {
+        let extra = match self.extra.as_mut() {
+            Some(extra) if Rc::strong_count(extra) == 1 => extra,
+            extra @ _ => {
+                self.extra =
+                    Some(Rc::new(extra.map(|extra| (**extra).clone()).unwrap_or_default()));
+                self.extra.as_mut().unwrap()
+            },
+        };
+
+        // We've ensured that particular arc, since we don't have weak refs, only strong.
+        let extra = Rc::get_mut(extra).unwrap();
+        extra.underline_color = color;
+    }
+
+    /// Underline color stored in this cell.
+    #[inline]
+    pub fn undreline_color(&self) -> Option<Color> {
+        self.extra.as_ref()?.underline_color
+    }
 }
 
+unsafe impl Send for Cell {}
+unsafe impl Sync for Cell {}
```

Comment:
This is wrong, however if we change clone impl so it clones the underlines extra storage and puts it into Rc it'll be fine? And we won't allow to modify the extra storage outside of Terminal.

Since the terminal is behind mutex in alacritty it's safe the way it's right now, however it won't be so if you move cell into another thread.